File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int findLeft(vector<int>& nums, int target) {
        int low = 0, high = nums.size() - 1, res = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target) {
                res = mid;
                high = mid - 1;  // Search left for the first occurrence
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return res;
    }

    int findRight(vector<int>& nums, int target) {
        int low = 0, high = nums.size() - 1, res = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] == target) {
                res = mid;
                low = mid + 1;  // Search right for the last occurrence
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return res;
    }

    vector<int> searchRange(vector<int>& nums, int target) {
        int leftIndex = findLeft(nums, target);
        if (leftIndex == -1) return {-1, -1};  // Target not found
        int rightIndex = findRight(nums, target);
        return {leftIndex, rightIndex};
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        # if target not in nums:
        #     return [-1,-1]
        # a=[]
        # for i in range(len(nums)):
        #     if nums[i]==target:
        #         a.append(i)
        # return [a[0],a[-1]] 

        def left(nums, target):
            low, high = 0, len(nums) - 1
            res = -1
            while low <= high:
                mid = (low + high) // 2
                if nums[mid] == target:
                    res = mid
                    high = mid - 1
                elif nums[mid] < target:
                    low = mid + 1
                else:
                    high = mid - 1
            return res
        
        def right(nums, target):
            low, high = 0, len(nums) - 1
            res = -1
            while low <= high:
                mid = (low + high) // 2
                if nums[mid] == target:
                    res = mid
                    low = mid + 1 
                elif nums[mid] < target:
                    low = mid + 1
                else:
                    high = mid - 1
            return res
        
        le = left(nums, target)
        if le == -1:
            return [-1, -1]
        ri = right(nums, target)
        return [le, ri] 